core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java

```diff
@@ -322,18 +319,19 @@ public Expression rewriteTrim(Trim node, Void context, ExpressionTreeRewriter<Vo
                     return coerceIfNecessary(node, mapped.get());
                 }
 
-                ImmutableList.Builder<Expression> expressions = ImmutableList.builder();
-                expressions.add(node.getTrimSource());
-                node.getTrimCharacter().ifPresent(trimCharacter -> expressions.add(new Cast(trimCharacter, toSqlType(CODE_POINTS))));
-
                 ResolvedFunction resolvedFunction = analysis.getResolvedFunction(node);
                 checkArgument(resolvedFunction != null, "Function has not been analyzed: %s", node);
 
-                FunctionCall functionCall = new FunctionCall(
+                Trim rewritten = treeRewriter.defaultRewrite(node, context);
+
+                ImmutableList.Builder<Expression> expressions = ImmutableList.builder();
+                expressions.add(rewritten.getTrimSource());
+                rewritten.getTrimCharacter().ifPresent(expressions::add);
+
+                return new FunctionCall(
                         node.getLocation().orElseThrow(),
                         resolvedFunction.toQualifiedName(),
                         expressions.build());
-                return coerceIfNecessary(functionCall, treeRewriter.defaultRewrite(functionCall, context));
```

Comment:
You still have to coerce the top level FunctionCall before you return it:
`return coerceIfNecessary(node, functionCall);`